# **Fake News Detection**

Do you trust all the news you hear from social media?

All news are not real, right?

How will you detect fake news?

The answer is **Python**. You will easily make a difference between **real and fake news**.

## **What is Fake News?**

A type of yellow journalism, fake news encapsulates pieces of news that may be hoaxes and is generally spread through social media and other online media. This is often done to further or impose certain ideas and is often achieved with political agendas. Such news items may contain false and/or exaggerated claims, and may end up being virtualized by algorithms, and users may end up in a filter bubble.

## **Detecting Fake News with Python**

To build a model to accurately classify a piece of news as REAL or FAKE.

## **About Detecting Fake News with Python**

This **python project** of **detecting** **fake news** deals with fake and real news. Using sklearn, we build a **Tfidf Vectorizer** on our dataset. Then, we initialize a **PassiveAggressive Classifier** and fit the model. In the end, the accuracy score and the **confusion matrix** tell us how well our **model fares**.

##**The fake news Dataset**

The dataset we’ll use for this python project- we’ll call it news.csv. This dataset has a shape of 7796×4. The first column identifies the news, the second and third are the title and text, and the fourth column has labels denoting whether the news is REAL or FAKE. The dataset takes up 29.2MB of space.

## **Code:**

In [2]:
# Make Necessary Imports
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
# Read the data
df = pd.read_csv("/content/drive/MyDrive/projects/Fake news Detection /news.csv")

In [4]:
df.shape

(6335, 4)

In [5]:
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [7]:
# Get labels from DataFrame (Output Variable = label)
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [8]:
# Split the dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size = 0.2, random_state = 7)

Let’s initialize a **TfidfVectorizer** with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a **TfidfVectorizer** turns a collection of raw documents into a matrix of TF-IDF features.

**Now, fit and transform the vectorizer on the train set, and transform the vectorizer on the test set**.

In [9]:
# Initialize a TfidfVectorizer
tfidf_vect = TfidfVectorizer(stop_words = "english", max_df = 0.7)

# Fit and transform train set, transform test set
tfidf_train = tfidf_vect.fit_transform(x_train)
tfidf_test = tfidf_vect.transform(x_test)

Next, we’ll initialize a **PassiveAggressiveClassifier**. This is. We’ll **fit** this on **tfidf_train and y_train**.

Then, we’ll **predict** on the **test set** from the **TfidfVectorizer** and calculate the **accuracy** with **accuracy_score()** from sklearn.metrics.

In [10]:
# Initialize a PassiveAggressiveClassifier
pac = PassiveAggressiveClassifier(max_iter = 50)
pac.fit(tfidf_train,y_train)

# Predict on the test set and calculate accuracy
y_pred = pac.predict(tfidf_test)
score = accuracy_score(y_test, y_pred)
print(f"Accuracy : {round(score*100,2)}%")

Accuracy : 92.58%


 **We got an accuracy of 92.58% with this model**. Finally, let’s print out a **confusion matrix** to **gain insight** into the **number of false and true negatives and positives**.

In [11]:
# Confusion matrix
confusion_matrix(y_test, y_pred, labels = ['FAKE', 'REAL'])

array([[586,  52],
       [ 42, 587]])

So with this model, we have **586 true positives, 587 true negatives, 42 false positives, and 52 false negatives.**

## **Conclusion:**

I hereby declared that my system detecting Fake and real news from a given dataset with 92.58% Accuracy Level. And also solve the issue of Yellow Journalism.